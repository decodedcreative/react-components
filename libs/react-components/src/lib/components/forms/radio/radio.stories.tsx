import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Radio } from './radio';

export default {
  title: 'Theme/Forms/Radio',
  component: Radio,
  parameters: {
    backgrounds: {
      default: 'Light',
      values: [
        { name: 'Light', value: '#ffffff' },
        { name: 'Dark', value: '#161b2f' },
      ],
    },
  },
  argTypes: {},
  args: {
    label: 'Male',
  },
} as ComponentMeta<typeof Radio>;

const Template: ComponentStory<typeof Radio> = (args) => <Radio {...args} />;

export const Default = Template.bind({});
Default.args = {};

export const Checked = Template.bind({});
Checked.args = {
  checked: true,
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
  checked: true,
};

export const Error = Template.bind({});
Error.args = {
  'aria-invalid': true,
  checked: true,
};
