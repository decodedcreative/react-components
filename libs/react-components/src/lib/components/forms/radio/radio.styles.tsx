import styled from 'styled-components';
import { RadioInputProps, RadioProps } from './radio';

export const Radio = styled('div')<RadioProps>`
  position: relative;
  display: flex;
`;

export const RadioLabel = styled('label')`
  color: ${(props) => props.theme.colors.color70};
  font-size: ${(props) => props.theme.typography.sizes.default};
  font-family: ${(props) => props.theme.typography.fontFamily};
  line-height: ${(props) => props.theme.typography.lineHeight};
  display: inline-flex;
  align-items: center;
  position: relative;
  z-index: -1;
`;

export const RadioBackground = styled('div')`
  background-color: ${(props) => props.theme.colors.color10};
  border: ${(props) => props.theme.general.borderWidth} solid
    ${(props) => props.theme.colors.color50};
  border-radius: 50%;
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 28px;
  height: 28px;
`;

export const RadioText = styled('span')`
  margin-left: ${(props) => props.theme.spacing.sm};
`;

export const RadioInput = styled('input').attrs<RadioInputProps>({
  type: 'Radio',
})<RadioInputProps>`
  position: absolute;
  opacity: 0;
  width: 24px;
  height: 24px;
  border-radius: ${(props) => props.theme.general.borderRadius};

  &:focus ~ ${RadioLabel} ${RadioBackground} {
    box-shadow: ${(props) => props.theme.general.focusBoxShadow};
    border-color: ${(props) => props.theme.colors.color70};
    outline-width: ${(props) => props.theme.general.focusWidth};
  }

  &:checked ~ ${RadioLabel} ${RadioBackground} {
    border-color: ${(props) => props.theme.colors.interactive};
  }

  &:checked:disabled ~ ${RadioLabel} ${RadioBackground} {
    border-color: ${(props) => props.theme.colors.color50};
  }

  &:disabled ~ ${RadioLabel} ${RadioBackground} {
    background-color: ${(props) => props.theme.colors.color30};
  }

  &:disabled ~ ${RadioLabel} ${RadioText} {
    color: ${(props) => props.theme.colors.color50};
  }

  &[aria-invalid='true'] ~ ${RadioLabel} ${RadioBackground} {
    border-color: ${(props) => props.theme.colors.error};
  }
`;

export const RadioIndicator = styled('div')`
  color: ${(props) => props.theme.colors.color10};
  transform: scale(0);
  transition: transform 0.05s ease-in;
  border-width: calc(16px / 2);
  border-style: solid;
  border-color: ${(props) => props.theme.colors.interactive};
  border-radius: 50%;

  ${RadioInput}:checked ~ ${RadioLabel} & {
    transform: scale(1);
  }
`;
