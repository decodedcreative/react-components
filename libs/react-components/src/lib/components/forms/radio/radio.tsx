import { ComponentPropsWithoutRef, forwardRef } from 'react';
import * as S from './radio.styles';

export interface RadioInputProps extends ComponentPropsWithoutRef<'input'> {}

export interface RadioProps extends RadioInputProps {
  label: string;
}

export const Radio = forwardRef<HTMLInputElement, RadioProps>(
  ({ label, ...props }, ref) => {
    const wrapperProps = {
      label,
    };

    const inputProps = {
      ...props,
    };

    return (
      <S.Radio {...wrapperProps}>
        <S.RadioInput {...inputProps} ref={ref} />
        <S.RadioLabel>
          <S.RadioBackground>
            <S.RadioIndicator />
          </S.RadioBackground>
          <S.RadioText>{label}</S.RadioText>
        </S.RadioLabel>
      </S.Radio>
    );
  }
);
