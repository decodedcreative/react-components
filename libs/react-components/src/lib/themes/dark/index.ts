import colors from './colors';
import components from './components';
import general from './general';
import spacing from './spacing';
import typography from './typography';

const darkTheme = {
  name: 'darkTheme',
  colors,
  components,
  general,
  spacing,
  typography,
};

export default darkTheme;
