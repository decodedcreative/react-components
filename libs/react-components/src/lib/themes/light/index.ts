import colors from './colors';
import components from './components';
import general from './general';
import typography from './typography';

const lightTheme = {
  name: 'lightTheme',
  colors,
  components,
  general,
  typography,
};

export default lightTheme;
