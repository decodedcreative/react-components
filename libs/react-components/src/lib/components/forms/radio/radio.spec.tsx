import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { lightTheme } from '../../../themes';
import { Radio, RadioProps } from './radio';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

const WithTheme = ({ label = 'Radio label', ...props }: RadioProps) => (
  <ThemeProvider theme={lightTheme}>
    <Radio label={label} {...props} />
  </ThemeProvider>
);

const createComponentJSX = ({ label, ...props }: RadioProps) => (
  <WithTheme label={label} {...props} />
);
const createComponentRender = ({ label, ...props }: RadioProps) =>
  render(<WithTheme label={label} {...props} />);

describe('Radio', () => {
  it('should render successfully', () => {
    const component = createComponentRender({} as RadioProps);
    expect(component).toBeTruthy();
  });

  it('should match the default snapshot', () => {
    const defaultRadio = createComponentJSX({} as RadioProps);
    const defaultRadioJSON = renderer.create(defaultRadio).toJSON();
    expect(defaultRadioJSON).toMatchSnapshot();
  });

  describe('Radio - States', () => {
    describe('Radio - States - Error', () => {
      it('should have an aria-invalid attribute on its input if in an error state', () => {
        const component = createComponentRender({
          label: 'Radio label',
          'aria-invalid': true,
          'data-testid': 'radio-input',
        } as RadioProps);
        const input = component.getByTestId('radio-input');
        expect(input).toHaveAttribute('aria-invalid');
      });
      it('should match the Radio With Error snapshot', () => {
        const errorRadio = createComponentJSX({
          'aria-invalid': true,
        } as RadioProps);
        const errorRadioJSON = renderer.create(errorRadio).toJSON();
        expect(errorRadioJSON).toMatchSnapshot();
      });
    });
    describe('Radio - States - Disabled', () => {
      it('should disable its input if a disabled prop is supplied', () => {
        const component = createComponentRender({
          label: 'Radio label',
          disabled: true,
          'data-testid': 'radio-input',
        } as RadioProps);
        const input = component.getByTestId('radio-input');
        expect(input).toBeDisabled();
      });
      it('should match the Disabled Radio snapshot', () => {
        const disabledRadio = createComponentJSX({
          disabled: true,
        } as RadioProps);
        const disabledRadioJSON = renderer.create(disabledRadio).toJSON();
        expect(disabledRadioJSON).toMatchSnapshot();
      });
    });
  });

  it('should display its label', () => {
    const component = createComponentRender({
      label: 'Radio label 2',
    } as RadioProps);
    expect(component.getByText('Radio label 2')).toBeTruthy();
  });
});
