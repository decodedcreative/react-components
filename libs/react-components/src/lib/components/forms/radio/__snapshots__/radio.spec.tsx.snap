// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Radio - States Radio - States - Disabled should match the Disabled Radio snapshot 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  color: #272727;
  font-size: 1rem;
  font-family: 'Source Sans Pro',sans-serif;
  line-height: 1.5;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  z-index: -1;
}

.c6 {
  background-color: #ffffff;
  border: 2px solid #8e8e8e;
  border-radius: 50%;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 28px;
  height: 28px;
}

.c10 {
  margin-left: 0.75rem;
}

.c2 {
  position: absolute;
  opacity: 0;
  width: 24px;
  height: 24px;
  border-radius: 5px;
}

.c2:focus ~ .c3 .c5 {
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
  border-color: #272727;
  outline-width: 2px;
}

.c2:checked ~ .c3 .c5 {
  border-color: #006de3;
}

.c2:checked:disabled ~ .c3 .c5 {
  border-color: #8e8e8e;
}

.c2:disabled ~ .c3 .c5 {
  background-color: #f6f6f6;
}

.c2:disabled ~ .c3 .c9 {
  color: #8e8e8e;
}

.c2[aria-invalid='true'] ~ .c3 .c5 {
  border-color: #e90014;
}

.c8 {
  color: #ffffff;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: -webkit-transform 0.05s ease-in;
  -webkit-transition: transform 0.05s ease-in;
  transition: transform 0.05s ease-in;
  border-width: calc(16px / 2);
  border-style: solid;
  border-color: #006de3;
  border-radius: 50%;
}

.c1:checked ~ .c3 .c7 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c1:disabled ~ .c3 .c7 {
  border-color: #8e8e8e;
}

.c1[aria-invalid="true"] ~ .c3 .c7 {
  border-color: #e90014;
}

<div
  className="c0"
  label="Radio label"
>
  <input
    className="c1 c2"
    disabled={true}
    type="Radio"
  />
  <label
    className="c3 c4"
  >
    <div
      className="c5 c6"
    >
      <div
        className="c7 c8"
      />
    </div>
    <span
      className="c9 c10"
    >
      Radio label
    </span>
  </label>
</div>
`;

exports[`Radio Radio - States Radio - States - Error should match the Radio With Error snapshot 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  color: #272727;
  font-size: 1rem;
  font-family: 'Source Sans Pro',sans-serif;
  line-height: 1.5;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  z-index: -1;
}

.c6 {
  background-color: #ffffff;
  border: 2px solid #8e8e8e;
  border-radius: 50%;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 28px;
  height: 28px;
}

.c10 {
  margin-left: 0.75rem;
}

.c2 {
  position: absolute;
  opacity: 0;
  width: 24px;
  height: 24px;
  border-radius: 5px;
}

.c2:focus ~ .c3 .c5 {
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
  border-color: #272727;
  outline-width: 2px;
}

.c2:checked ~ .c3 .c5 {
  border-color: #006de3;
}

.c2:checked:disabled ~ .c3 .c5 {
  border-color: #8e8e8e;
}

.c2:disabled ~ .c3 .c5 {
  background-color: #f6f6f6;
}

.c2:disabled ~ .c3 .c9 {
  color: #8e8e8e;
}

.c2[aria-invalid='true'] ~ .c3 .c5 {
  border-color: #e90014;
}

.c8 {
  color: #ffffff;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: -webkit-transform 0.05s ease-in;
  -webkit-transition: transform 0.05s ease-in;
  transition: transform 0.05s ease-in;
  border-width: calc(16px / 2);
  border-style: solid;
  border-color: #006de3;
  border-radius: 50%;
}

.c1:checked ~ .c3 .c7 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c1:disabled ~ .c3 .c7 {
  border-color: #8e8e8e;
}

.c1[aria-invalid="true"] ~ .c3 .c7 {
  border-color: #e90014;
}

<div
  className="c0"
  label="Radio label"
>
  <input
    aria-invalid={true}
    className="c1 c2"
    type="Radio"
  />
  <label
    className="c3 c4"
  >
    <div
      className="c5 c6"
    >
      <div
        className="c7 c8"
      />
    </div>
    <span
      className="c9 c10"
    >
      Radio label
    </span>
  </label>
</div>
`;

exports[`Radio should match the default snapshot 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  color: #272727;
  font-size: 1rem;
  font-family: 'Source Sans Pro',sans-serif;
  line-height: 1.5;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  z-index: -1;
}

.c6 {
  background-color: #ffffff;
  border: 2px solid #8e8e8e;
  border-radius: 50%;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 28px;
  height: 28px;
}

.c10 {
  margin-left: 0.75rem;
}

.c2 {
  position: absolute;
  opacity: 0;
  width: 24px;
  height: 24px;
  border-radius: 5px;
}

.c2:focus ~ .c3 .c5 {
  box-shadow: 0 0 0 2px #ffffff,0 0 0 4px #006de3;
  border-color: #272727;
  outline-width: 2px;
}

.c2:checked ~ .c3 .c5 {
  border-color: #006de3;
}

.c2:checked:disabled ~ .c3 .c5 {
  border-color: #8e8e8e;
}

.c2:disabled ~ .c3 .c5 {
  background-color: #f6f6f6;
}

.c2:disabled ~ .c3 .c9 {
  color: #8e8e8e;
}

.c2[aria-invalid='true'] ~ .c3 .c5 {
  border-color: #e90014;
}

.c8 {
  color: #ffffff;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition: -webkit-transform 0.05s ease-in;
  -webkit-transition: transform 0.05s ease-in;
  transition: transform 0.05s ease-in;
  border-width: calc(16px / 2);
  border-style: solid;
  border-color: #006de3;
  border-radius: 50%;
}

.c1:checked ~ .c3 .c7 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c1:disabled ~ .c3 .c7 {
  border-color: #8e8e8e;
}

.c1[aria-invalid="true"] ~ .c3 .c7 {
  border-color: #e90014;
}

<div
  className="c0"
  label="Radio label"
>
  <input
    className="c1 c2"
    type="Radio"
  />
  <label
    className="c3 c4"
  >
    <div
      className="c5 c6"
    >
      <div
        className="c7 c8"
      />
    </div>
    <span
      className="c9 c10"
    >
      Radio label
    </span>
  </label>
</div>
`;
